component/encryptor/encryptor.go

```diff
@@ -0,0 +1,382 @@
+		// Pad the data to the block size
+		paddingLength = int64(blockSize) - int64(len(data))
+		data = append(data, make([]byte, paddingLength)...)
+	}
+	encryptedChunk, nonce, err := EncryptChunk(data, encryptionKey)
+	if err != nil {
+		log.Err("Encryptor: Error encrypting data: %s", err.Error())
+		return err
+	}
+
+	encryptedChunkOffset := int64(blockId) * (int64(blockSize) + int64(MetaSize))
+	// Write the combined nonce and encrypted chunk
+	n, err := encryptedFile.WriteAt(append(nonce, encryptedChunk...), encryptedChunkOffset)
+	log.Debug("Encryptor:: encryptWriteBlock: writing %d bytes to encrypted file: %s at offset %d", n, name, encryptedChunkOffset)
+	if err != nil {
+		log.Err("Encryptor: Error writing encrypted chunk to encrypted file: %s at offset %d : size of data %d", err.Error(), encryptedChunkOffset, n)
+		return err
+	}
+
+	lastchunkMeta.Lock()
+	defer lastchunkMeta.Unlock()
+	if int64(blockId) > lastchunkMeta.farthestBlockSeen {
+		lastchunkMeta.farthestBlockSeen = int64(blockId)
+	}
+	lastchunkMeta.paddingLength = paddingLength
+	log.Debug("Encryptor:: encryptWriteBlock: endBlockIndex %d, paddingLength %d", lastchunkMeta.farthestBlockSeen, lastchunkMeta.paddingLength)
+	return nil
+}
+
+func (e *Encryptor) ReadInBuffer(options internal.ReadInBufferOptions) (length int, err error) {
+	log.Trace("Encryptor::ReadInBuffer : Read %s from %d offset", options.Handle.Path, options.Offset)
+
+	if options.Offset > atomic.LoadInt64(&options.Handle.Size) {
+		return 0, syscall.ERANGE
+	}
+
+	var dataLen int64 = int64(len(options.Data))
+	if atomic.LoadInt64(&options.Handle.Size) < (options.Offset + int64(len(options.Data))) {
+		dataLen = options.Handle.Size - options.Offset
+	}
+
+	if dataLen == 0 {
+		return 0, nil
+	}
+
+	err = readAndDecryptBlock(options.Handle.Path, options.Offset, options.Data, e.blockSize, e.encryptionKey, e.mountPointCipher)
+	if err != nil {
+		log.Err("Encryptor::ReadInBuffer : Failed to read %s [%s] from offset %d", options.Handle.Path, err.Error(), options.Offset)
+		return 0, err
+	}
+
+	length = int(dataLen)
+	return
+}
+
+func readAndDecryptBlock(name string, offset int64, data []byte, blockSize uint64, encryptionKey []byte, mountPointCipher string) error {
+	log.Debug("Encryptor::ReadAndDecryptBlock : Read %s from %d offset for data size %d", name, offset, len(data))
+
+	fileHandle, err := os.OpenFile(mountPointCipher+name, os.O_RDONLY, 0666)
+	if err != nil {
+		log.Err("Encryptor::ReadAndDecryptBlock : Error opening encrypted file: %s", err.Error())
+		return err
+	}
+	defer fileHandle.Close()
+	chunkIndex := offset / int64(blockSize)
+	encryptedChunkOffset := chunkIndex * (int64(blockSize) + MetaSize)
+	encryptedChunk := make([]byte, blockSize+MetaSize)
+
+	n, err := fileHandle.ReadAt(encryptedChunk, encryptedChunkOffset)
+	if err != nil {
+		log.Err("Encryptor: Error reading encrypted file: %s", err.Error())
+		return err
+	}
+
+	log.Debug("Encryptor: Read %d bytes from encrypted file", n)
+
```

Comment:
When will logs like these be useful? We can be mindful about Debug level logs also and not spew too much if its really just noise.

Reply:
This was to ensure bytes read from file is captured.

Reply:
Remove now if you are done with it?

Reply:
shouldn't we keep these debug logs in case there is a failure so we run the flow again with debug level logs .?